### PR TITLE
Implementation of PAM encoder/decoder

### PR DESCRIPTION
**Pull request by DmtKats**
**Merge with**: https://github.com/opencv/opencv_extra/pull/306
**Original PR**: https://github.com/opencv/opencv/pull/6912

This pull request implements the PAM image format as defined at:
http://netpbm.sourceforge.net/doc/pam.html

The PAM format provides a generic means for storing 2 dimensional information.
This is useful for opencv since there are cases where data gets translated into
non standardized formats, which makes it difficult to store and load this information.

I have reopened the PR for opencv and opencv_extra as per the suggestion of alalek. This was done since the testing system ignores branches with names like "master" and "2.4"

### CODE REVIEW

**Comment by DmtKats** [comment]
As with the previous PR, the implemented tests will fail unless the PAM image in extras is merged first.

**Comment by alalek** [comment]
OpenCV extra patch is "merged" for testing now.

Implemented new tests work [fine](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64_no_opt/builds/4756/steps/test_imgcodecs/logs/stdio):
```
[----------] 2 tests from Imgcodecs_Pam
[ RUN      ] Imgcodecs_Pam.encode_empty
[       OK ] Imgcodecs_Pam.encode_empty (0 ms)
[ RUN      ] Imgcodecs_Pam.readwrite
[       OK ] Imgcodecs_Pam.readwrite (26 ms)
[----------] 2 tests from Imgcodecs_Pam (26 ms total)
```

But modified `Imgcodecs_Image.write_imageseq` test is failed.

File: modules/imgcodecs/test/test_grfmt.cpp [hunk 1]
```diff
@@ -1017,3 +1019,35 @@ TEST(Imgcodecs_Hdr, regression)
         ASSERT_FALSE(max > DBL_EPSILON);
     }
 }
+
+TEST(Imgcodecs_Pam, encode_empty)
+{
+    cv::Mat img;
+    std::vector<uchar> jpegImg;
+
+    ASSERT_THROW(cv::imencode(".jpg", img, jpegImg), cv::Exception);
```
Comment (alalek) [inline modules/imgcodecs/test/test_grfmt.cpp]: "jpg/jpeg" -> "pam"

**Comment by DmtKats** [comment]
Ok, so I found a bug with the signature check for gdal. I added a patch to fix it and the unittests pass now.

**Comment by vpisarev** [comment]
@DmtKats, in your PR I see patches from other PRs, like HOG optimization etc. Please, submit fresh PR with just your changes included.

**Comment by DmtKats** [comment]
I removed all the other patches and the build has succeded

File: modules/imgcodecs/src/grfmt_gdal.cpp [hunk 2]
```diff
@@ -559,14 +559,18 @@ bool GdalDecoder::checkSignature( const String& signature )const{
 
 
     // look for NITF
-    std::string str = signature.c_str();
```
Comment (alalek) [inline modules/imgcodecs/src/grfmt_gdal.cpp]: This [should](https://github.com/opencv/opencv/blob/28db4a220741b33ba59a30fac6dd8e470f444cf6/modules/core/include/opencv2/core/cvstd.inl.hpp#L115-L119) work fine with null terminated symbols:
```
std::string str(signature);
```

**Comment by DmtKats** [comment]
I implemented the change to string and it passes

**Comment by alalek** [comment]
@DmtKats Thanks! Looks good to me!
:+1:

**Pull request merged**